<a href="https://colab.research.google.com/github/sas4wy/assignment2/blob/main/Assignment2Updated.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/DS3001/assignment2`

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about? **This paper is about a structure/framework that makes it easy to "tidy" messy datasets.**
  2. Read the introduction. What is the "tidy data standard" intended to accomplish? **The "tidy data standard" is intended to simplify data organization and initial data analysis by already having established principles for cleaning.**
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general." **The first sentence indicates that tidy data is standardized and can be linked with each other is some way, whereas messy data does not have this trait, and is in a sense, "all over the place." The latter sentence indicates that when looking at a dtataset, what is a variable and what is an observation is easily distinguishable, but when defining the two more generally, they often overlap.**
  4. Read Section 2.2. How does Wickham define values, variables, and observations? **Wickam defines values as numbers or strings depending on the dataset, variables as a collection of  values that measure a shared attribute, and observations as values measured on the same unit.**
  5. How is "Tidy Data" defined in section 2.3? **"Tidy data" is a standard way of mapping the meaning of a dataset to its structure. Each variable forms a column, easy observation forms a row, each observational unit forms a table.**
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset? **The five most common problems with messy datasets are: column headers as values and not variable names, multiple varibles stored in one column, variables in both rows and columns, multiple types of observational units in the same table or a single obervational unit in multiple tables. In table four, the column headers are numbers/values and not actual titles.**
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?  **Table 11 is messy because it has variables across rows and columns, not only individual columns. Table 12 is "molten" because the values of a column are variables.**
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?  **The "chicken-and-egg" problem is that tidy data is only as useful as the tools that work with it so in essence, they are linked. Wickham hopes that other frameworks for  issues associated with data cleaning (i.e missing values) will emerge making these tasks easier as well.**

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?) **To clean the data, I first wanted to see how "dirty" it was, and get familiar with it. I also wanted to see if the values were classifed as numeric and not objects. I then eliminated commas, and converted the values from objects to numeric values. After this, I did end up with missing values (181).**
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make. **To clean this data I also used the .unique(), .value_counts(), and .describe() in order to see how "dirty" it was. I then saw many "invalids" and checked the data itself to see what they could be. I then renamed the variable to "Possibly Attacked" and combined the "Under Investigation" variable as well. I then combined the Boating, Boat, and Watercraft varibles under one variable: Boating. Others, such as "Questionable" or "Unverified" were listed as missing.**
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.) **A look at the codebook indicates that those listed as 4 or 9 were either missing, released, or etc, indicating that these could be classified as missing data, thus, these were updated to nans.**

In [ ]:
! git clone https://www.github.com/sas4wy/assignment2

fatal: destination path 'assignment2' already exists and is not an empty directory.


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('./assignment2/data/airbnb_hw.csv')

In [ ]:
var = 'Price'
print(df[var].unique(),'\n')
print(df[var].describe(),'\n')

['145' '37' '28' '199' '549' '149' '250' '90' '270' '290' '170' '59' '49'
 '68' '285' '75' '100' '150' '700' '125' '175' '40' '89' '95' '99' '499'
 '120' '79' '110' '180' '143' '230' '350' '135' '85' '60' '70' '55' '44'
 '200' '165' '115' '74' '84' '129' '50' '185' '80' '190' '140' '45' '65'
 '225' '600' '109' '1,990' '73' '240' '72' '105' '155' '160' '42' '132'
 '117' '295' '280' '159' '107' '69' '239' '220' '399' '130' '375' '585'
 '275' '139' '260' '35' '133' '300' '289' '179' '98' '195' '29' '27' '39'
 '249' '192' '142' '169' '1,000' '131' '138' '113' '122' '329' '101' '475'
 '238' '272' '308' '126' '235' '315' '248' '128' '56' '207' '450' '215'
 '210' '385' '445' '136' '247' '118' '77' '76' '92' '198' '205' '299'
 '222' '245' '104' '153' '349' '114' '320' '292' '226' '420' '500' '325'
 '307' '78' '265' '108' '123' '189' '32' '58' '86' '219' '800' '335' '63'
 '229' '425' '67' '87' '1,200' '158' '650' '234' '310' '695' '400' '166'
 '119' '62' '168' '340' '479' '43' '395' '144' '52' 

In [ ]:
df[var].str.replace(",","")
df[var] = pd.to_numeric(df[var], errors='coerce')

In [ ]:
print(df[var].describe(),'\n')
df[var+'_nan'] = df[var].isnull()
print('Total Missings: \n', sum(df[var+'_nan']),'\n')

count    30297.000000
mean       153.357032
std        108.895698
min         10.000000
25%         80.000000
50%        125.000000
75%        192.000000
max        999.000000
Name: Price, dtype: float64 

Total Missings: 
 181 



In [ ]:
df = pd.read_csv('./assignment2/data/sharks.csv')

In [ ]:
var2 = 'Type'
print(df[var2].unique(),'\n')
print(df[var2].value_counts(), '\n')
print(df[var2].describe(),'\n')

['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Unconfirmed'
 'Unverified' 'Invalid' 'Under investigation' 'Boating' 'Sea Disaster' nan
 'Boat' 'Boatomg'] 

Unprovoked             4716
Provoked                593
Invalid                 552
Sea Disaster            239
Watercraft              142
Boat                    109
Boating                  92
Questionable             10
Unconfirmed               1
Unverified                1
Under investigation       1
Boatomg                   1
Name: Type, dtype: int64 

count           6457
unique            12
top       Unprovoked
freq            4716
Name: Type, dtype: object 



In [ ]:
df['Type'] = df['Type'].replace({
    'Invalid': 'Possibly Attacked',
    'Under investigation': 'Possibly Attacked',
    'Unverified': 'Missing',
    'Unconfirmed': 'Missing',
    'Questionable': 'Missing',
    'Boatomg': 'Boating',
    'Watercraft': 'Boating',
    'Boat': 'Boating'
})

In [ ]:
print(df[var2].value_counts(), '\n')

Unprovoked           4716
Provoked              593
Possibly Attacked     553
Boating               344
Sea Disaster          239
Missing                12
Name: Type, dtype: int64 



In [ ]:
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
df = pd.read_csv(url,low_memory=False)

In [ ]:
var3 = 'WhetherDefendantWasReleasedPretrial'
print(df[var3].unique(),'\n')
print(df[var3].value_counts(), '\n')
print(df[var3].describe(),'\n')

[9 0 1] 

1    19154
0     3801
9       31
Name: WhetherDefendantWasReleasedPretrial, dtype: int64 

count    22986.000000
mean         0.845428
std          0.477276
min          0.000000
25%          1.000000
50%          1.000000
75%          1.000000
max          9.000000
Name: WhetherDefendantWasReleasedPretrial, dtype: float64 



In [ ]:
df[var3] = df[var3].replace(9, np.nan)
print(df[var3].value_counts())

1.0    19154
0.0     3801
Name: WhetherDefendantWasReleasedPretrial, dtype: int64


In [ ]:
var4 = 'ImposedSentenceAllChargeInContactEvent'
print(df[var4].unique(),'\n')
print(df[var4].describe(),'\n')
print(df[var4].value_counts(), '\n')

[' ' '60' '12' '.985626283367556' '36' '6' '24' '5.91375770020534' '120'
 '72' '11.9917864476386' '0' '2.95687885010267' '84' '108' '300' '240'
 '180' '4' '96' '2' '54' '.328542094455852' '44' '5' '115' '132' '48'
 '258' '34' '76' '.164271047227926' '.131416837782341' '111' '9' '3'
 '1.97125256673511' '36.9856262833676' '.0657084188911704'
 '35.4928131416838' '106.492813141684' '8' '35' '18.3141683778234' '480'
 '32' '93' '234' '732' '1.16427104722793' '4.6570841889117' '21' '7'
 '4.49281314168378' '18' '600' '43.1642710472279' '179' '52' '30' '20'
 '192' '702' '14' '55' '53' '11.9055441478439' '114' '35.0061601642711'
 '68' '.657084188911704' '46.6242299794661' '102' '65' '200' '57'
 '24.3285420944559' '12.1642710472279' '117' '81.4928131416838'
 '22.4928131416838' '1980' '3.6570841889117' '56' '10' '2.79260780287474'
 '1' '47' '22' '1500' '40' '284' '11' '118' '42' '162' '156'
 '47.2956878850103' '105' '51' '246' '29' '75' '324' '360'
 '34.4804928131417' '120.328542094456' '59.926078

In [ ]:
pd.crosstab(df['ImposedSentenceAllChargeInContactEvent'], df['SentenceTypeAllChargesAtConvictionInContactEvent'])

SentenceTypeAllChargesAtConvictionInContactEvent  0   1  2     4    9
ImposedSentenceAllChargeInContactEvent                               
                                                  0   0  0  8779  274
.0328542094455852                                 0   6  0     0    0
.0657084188911704                                 3  21  0     0    0
.0985626283367557                                 2   6  0     0    0
.131416837782341                                  4  10  0     0    0
...                                              ..  .. ..   ...  ...
98                                                0   0  1     0    0
98.95687885                                       0   0  1     0    0
99                                                0   1  0     0    0
99.3285420944559                                  0   1  0     0    0
99.4928131416838                                  0   1  0     0    0

[484 rows x 5 columns]

In [ ]:
df.loc[df['SentenceTypeAllChargesAtConvictionInContactEvent'] == (4 or 9), 'ImposedSentenceAllChargeInContactEvent'] = np.nan

In [ ]:
df[var4] = df[var4].replace(' ',np.nan)

In [ ]:
print(df[var4].unique(),'\n')
print(df[var4].describe(),'\n')
print(df[var4].value_counts(), '\n')

[nan '60' '12' '.985626283367556' '36' '6' '24' '5.91375770020534' '120'
 '72' '11.9917864476386' '0' '2.95687885010267' '84' '108' '300' '240'
 '180' '4' '96' '2' '54' '.328542094455852' '44' '5' '115' '132' '48'
 '258' '34' '76' '.164271047227926' '.131416837782341' '111' '9' '3'
 '1.97125256673511' '36.9856262833676' '.0657084188911704'
 '35.4928131416838' '106.492813141684' '8' '35' '18.3141683778234' '480'
 '32' '93' '234' '732' '1.16427104722793' '4.6570841889117' '21' '7'
 '4.49281314168378' '18' '600' '43.1642710472279' '179' '52' '30' '20'
 '192' '702' '14' '55' '53' '11.9055441478439' '114' '35.0061601642711'
 '68' '.657084188911704' '46.6242299794661' '102' '65' '200' '57'
 '24.3285420944559' '12.1642710472279' '117' '81.4928131416838'
 '22.4928131416838' '1980' '3.6570841889117' '56' '10' '2.79260780287474'
 '1' '47' '22' '1500' '40' '284' '11' '118' '42' '162' '156'
 '47.2956878850103' '105' '51' '246' '29' '75' '324' '360'
 '34.4804928131417' '120.328542094456' '59.926078

**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_hw.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race? **The most recent US Census gathered data on race based on self-reporting using a questionnaire.**

2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter? **We gather these data in order to get a sense of what America's demographic looks like, but data such as this is important in politics and society as it can help with making policy with issues regarding race and civil rights. Data quality matters because the most accurate data will help make the most informed decision about these issues.**

3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data? **One of the things that the Census did well was be accessible. The questionnaire can be accessed via the internet or by mail and etc. However, the U.S Census does not take into account gender identity or sexual orientation, which I would consider to be very important in our society today - it seems as though they ignore it. In the future, I think that we should aim to be more inclusive to the aspects above, as well as possibly religion. I think that these practices would be very beneficial if adopted much more widely - it would reflect the nuance and complexities that most people have.**
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices. **As I stated above, they indicated that they do not ask questions about gender or sex at birth - individuals can only choose male or female. Similar to the prior question, this choice ignores the nuance of inidividuals and is an incredibly important aspect of the data, especially for those who do not identify with either of those choices.**
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why? **When it comes to cleaning data, it may be a bit harder, as there may be more inconsistencies and errors since it cannot be captured with simple answer choices or categories. With missing values, it can be hard to replace to them, so they could potentially be eliminated from the set. I think that people may start adapting more effective data cleaning strategies to adapt to growing complexities.**
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have? **The major concerns are consent and privacy. Imputing values for characteristics that are unique to the indivudual that are considered "protected" is a major privacy violation, especially for those not giving consent. If someone chooses to withold that information, that is completely their choice.**

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.